app/src/main/java/com/guru/bluetooth/MainActivity.kt

```diff
@@ -1,194 +1,248 @@
+    private var bluetoothAdapter: BluetoothAdapter? = null
+    private val permissionsHelper = PermissionsHelper(this)
+    private lateinit var pairedDevices: Set<BluetoothDevice>
+    private lateinit var discoveredDevices: ArrayList<BluetoothDevice>
+    private var isInScanningMode: Boolean = false
     private val requestEnableBtLauncher = registerForActivityResult(
         ActivityResultContracts.StartActivityForResult()
     ) { result ->
         if (result.resultCode != RESULT_OK) {
-            Toast.makeText(this, "Bluetooth must be enabled", Toast.LENGTH_SHORT).show()
-            finish()
-        }
-    }
-
-    private val requestDiscoverableLauncher = registerForActivityResult(
-        ActivityResultContracts.StartActivityForResult()
-    ) { result ->
-        if (result.resultCode != RESULT_OK) {
-            Toast.makeText(this, "Device must be discoverable", Toast.LENGTH_SHORT).show()
+            showToast("Bluetooth must be enabled")
             finish()
-        } else {
-            startServer()
         }
     }
 
-    private val requestPermissionLauncher =
-        registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
-            if (isGranted) {
-                scanForDevices(this)
-            } else {
-                Toast.makeText(this, "Permission denied", Toast.LENGTH_SHORT).show()
-                finish()
-            }
-        }
-
-    private val deviceClickListener = { device: BluetoothDevice ->
-        bluetoothDevice = device
-        connectToDevice(device)
-    }
-
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
-        setupViews()
+        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
+
+        permissionsHelper.requestAccessCoarseLocationPermissionIfNotGranted()
+        permissionsHelper.requestReadExternalStoragePermissionIfNotGranted()
+
+        discoveredDevices = ArrayList()
+
+        val deviceFilter = IntentFilter(BluetoothDevice.ACTION_FOUND)
+        registerReceiver(receiver, deviceFilter)
+
+        val deviceNameFilter = IntentFilter(BluetoothDevice.ACTION_NAME_CHANGED)
+        registerReceiver(nameReceiver, deviceNameFilter)
+
+        initListener()
         enableBluetooth()
+        launchEnableBluetoothActivity()
+        launchEnableDiscoverableActivity()
+    }
+
+    private fun initListener() {
+        binding.mainEnterZone.setOnClickListener { enterScanningMode() }
+        binding.mainRefreshUserList.setOnClickListener { refreshList() }
     }
 
-    private fun setupViews() {
-        binding.selectFileButton.setOnClickListener {
-            selectFile()
+    private fun enableBluetooth() {
+        if (bluetoothAdapter == null) {
+            showToast("This device does not support Bluetooth")
+            this.finish()
+            return
```

Comment:
do we need the return statement if the activity is already destroyed using  this.finish() at line no. 72 ?

Reply:
don't need to use a return